src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -393,27 +394,125 @@ private Task StartAsyncCore(TransferFormat transferFormat)
+                                            {
+                                                break;
+                                            }
+                                        }
+                                        break;
+                                }
+                                break;
+                            case JsonToken.EndObject:
+                                completed = true;
+                                break;
+                        }
+                    } while (!completed && JsonUtils.CheckRead(reader));
+
+                    return new NegotiationResponse
+                    {
+                        ConnectionId = connectionId,
+                        AvailableTransports = availableTransports
+                    };
                 }
             }
+            catch (Exception ex)
+            {
+                throw new FormatException("Invalid negotiation response received.", ex);
+            }
+        }
+
+        private static AvailableTransport ParseAvailableTransport(JsonTextReader reader)
+        {
+            AvailableTransport availableTransport = new AvailableTransport();
 
-            if (negotiationResponse == null)
+            while (JsonUtils.CheckRead(reader))
             {
-                throw new FormatException("Invalid negotiation response received.");
+                switch (reader.TokenType)
+                {
+                    case JsonToken.PropertyName:
+                        string memberName = reader.Value.ToString();
+
+                        switch (memberName)
+                        {
+                            case "transport":
+                                availableTransport.Transport = JsonUtils.ReadAsString(reader, "transport");
+                                break;
+                            case "transferFormats":
+                                JsonUtils.CheckRead(reader);
+                                JsonUtils.EnsureArrayStart(reader);
+
+                                bool completed = false;
+                                availableTransport.TransferFormats = new List<string>();
+                                while (!completed && JsonUtils.CheckRead(reader))
+                                {
+                                    switch (reader.TokenType)
+                                    {
+                                        case JsonToken.String:
+                                            availableTransport.TransferFormats.Add(reader.Value.ToString());
+                                            break;
+                                        case JsonToken.EndArray:
+                                            completed = true;
+                                            break;
+                                        default:
+                                            throw new InvalidOperationException("Unexpected token.");
```

Comment:
TODO: Improve error messages

Reply:
Let's make the exception type consistent too. We use `InvalidDataException` and `FormatException` somewhat haphazardly currently. We should consolidate to `InvalidDataException` (you don't necessarily have to go to code you aren't touching in this PR, but for the code you're already touching, let's make it consistent)